clib/memprof_coq.std.ml

```diff
@@ -0,0 +1,144 @@
+let is_interrupted _ = false [@@inline]
+
+module Resource_bind = struct
+  let ( let& ) f scope = f ~scope
+end
+
+module Mutex_aux = struct
+  external unlock: Mutex.t -> unit = "caml_mutex_unlock"
+
+  (* Critical sections :
+     - Mutex.lock does not poll on leaving the blocking section
+       since 4.12.
+     - Never inline, to avoid theoretically-possible reorderings with
+       flambda.
+     - Inline the call to Mutex.unlock to avoid polling in bytecode.
+       (workaround to the lack of masking) *)
+  let[@inline never] with_lock m ~scope =
+    let () = Mutex.lock m (* BEGIN ATOMIC *) in
+    match (* END ATOMIC *) scope () with
```

Comment:
what do these atomic comments mean?

Reply:
Again coming from `memprof-limits`, I guess it is related to the low-level semantics of the OCaml runtime.

Reply:
See also #14046

Reply:
https://github.com/coq/coq/pull/14046#discussion_r605675906

Reply:
Note that the async-safe Mutex.protect is now also available upstream (since 4.14 I think).